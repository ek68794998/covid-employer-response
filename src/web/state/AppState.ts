import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}