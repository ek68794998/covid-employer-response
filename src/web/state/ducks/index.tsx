import { combineReducers } from "redux";

import { getLocalizedStrings } from "./localization/reducers";

export default combineReducers({
	strings: getLocalizedStrings,
});