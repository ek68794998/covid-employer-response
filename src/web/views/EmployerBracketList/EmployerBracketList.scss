.EmployerBracketList__Item {
	margin-bottom: 12px;
}