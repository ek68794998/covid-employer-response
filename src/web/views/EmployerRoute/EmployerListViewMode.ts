// For maintainability of union types, do not require a typedef.
// eslint-disable-next-line
export const EmployerListViewModeValues = [
	"cards",
	"rows",
] as const;

export type EmployerListViewMode = typeof EmployerListViewModeValues[number];
