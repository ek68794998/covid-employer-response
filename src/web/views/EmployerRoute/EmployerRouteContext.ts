/* eslint-disable unused-imports/no-unused-vars-ts */
/* Methods declared in the interface must be used in the context default value, but don't use their arguments. */

import React from "react";

import { EmployerListSearchFilter } from "../EmployerListSearch/EmployerListSearchFilter";

type ContextDataSetter<T> = (value: T) => void;

const getDefaultContextDataSetter = <T>(): ContextDataSetter<T> =>
	(value: T): void => (((): void => { /* Noop. */ }) as any)(value);

export const DefaultContextData: EmployerRouteContextData = {
	pageIndex: 1,
	searchFilters: new EmployerListSearchFilter(),
	setPageIndex: getDefaultContextDataSetter<number>(),
	setSearchFilters: getDefaultContextDataSetter<EmployerListSearchFilter>(),
};

export interface EmployerRouteContextData {
	pageIndex: number;

	searchFilters: EmployerListSearchFilter;

	setPageIndex: ContextDataSetter<number>;

	setSearchFilters: ContextDataSetter<EmployerListSearchFilter>;
}

export const EmployerRouteContext: React.Context<EmployerRouteContextData> =
	React.createContext(DefaultContextData);
