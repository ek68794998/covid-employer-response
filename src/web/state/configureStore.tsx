import { AnyAction, applyMiddleware, compose, createStore, Store } from "redux";
import thunk from "redux-thunk";

import { Production } from "../../common/constants/EnvironmentConstants";

import { AppState } from "./AppState";
import rootReducer from "./ducks";

const configureStore = (preloadedState: Partial<AppState>): Store<AppState, AnyAction> => {
	const isProd: boolean = process.env.NODE_ENV === Production;
	const composeEnhancers: any =
		(!isProd && typeof window !== "undefined" && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

	const store: Store<AppState, AnyAction> = createStore(
		rootReducer,
		preloadedState,
		composeEnhancers(applyMiddleware(thunk)),
	);

	if (module.hot) {
		// Enable Webpack hot module replacement for reducers
		module.hot.accept("./ducks", () => {
			const nextRootReducer: any = require("./ducks").default;
			store.replaceReducer(nextRootReducer);
		});
	}

	return store;
};

export default configureStore;