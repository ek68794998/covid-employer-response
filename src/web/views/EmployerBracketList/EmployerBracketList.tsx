import React from "react";

import { EmployerRecordMetadata } from "../../../common/EmployerRecordMetadata";

import EmployerListItem, { EmployerMetric } from "../EmployerListItem/EmployerListItem";

import "./EmployerBracketList.scss";

interface Props {
	employers: EmployerRecordMetadata[];

	metric: EmployerMetric;

	title: string;
}

const EmployerBracketList: React.FC<Props> = (props: Props): React.ReactElement => {
	const { employers, metric, title } = props;

	return (
		<div className="EmployerBracketList__Container">
			<h2>{title}</h2>
			<div className="EmployerBracketList__Grid">
				{employers.map(
					(e: EmployerRecordMetadata, i: number): JSX.Element => (
						<div className="EmployerBracketList__Item" key={`${i}-${e.id}`}>
							<EmployerListItem
								employerId={e.id}
								metric={metric}
								showDetails={false}
							/>
						</div>
					))}
			</div>
		</div>
	);
};

export default EmployerBracketList;
