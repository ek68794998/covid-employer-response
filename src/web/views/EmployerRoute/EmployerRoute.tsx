import React, { useEffect, useState } from "react";
import { RouteComponentProps } from "react-router-dom";

import EmployerBracketList from "../EmployerBracketList/EmployerBracketList";
import EmployerListPage from "../EmployerListPage/EmployerListPage";
import EmployerPage from "../EmployerPage/EmployerPage";

import { EmployerListViewMode, EmployerListViewModeValues } from "./EmployerListViewMode";
import { DefaultContextData, EmployerRouteContext, EmployerRouteContextData } from "./EmployerRouteContext";

export const RankingsRoute: string = "ranking";
export const RecentsRoute: string = "recent";

interface Params {
	id?: string;
}

type Props = RouteComponentProps<Params>;

const canUseSessionStorage: boolean = typeof sessionStorage !== "undefined";

const EmployerRoute: React.FC<Props> = (props: Props): React.ReactElement => {
	const [ listViewMode, setListViewMode ] = useState(DefaultContextData.listViewMode);
	const [ pageIndex, setPageIndex ] = useState(DefaultContextData.pageIndex);
	const [ searchFilters, setSearchFilters ] = useState(DefaultContextData.searchFilters);

	const contextData: EmployerRouteContextData = {
		listViewMode,
		pageIndex,
		searchFilters,
		setListViewMode,
		setPageIndex,
		setSearchFilters,
	};

	if (canUseSessionStorage) {
		useEffect(
			(): void => {
				const sessionListViewMode: string | null =
					canUseSessionStorage ? sessionStorage.getItem("EmployerRoute.listViewMode") : null;

				if (EmployerListViewModeValues.indexOf(sessionListViewMode as any) >= 0) {
					setListViewMode(sessionListViewMode as EmployerListViewMode);
				}
			},
			[]);

		useEffect(
			(): void => {
				sessionStorage.setItem("EmployerRoute.listViewMode", listViewMode);
			},
			[ listViewMode ]);
	}

	const employerId: string | undefined = props.match.params.id;

	let employerRoutedComponent: JSX.Element;

	if (employerId) {
		if (employerId === RecentsRoute) {
			employerRoutedComponent = <EmployerBracketList mode="recent" />;
		} else if (employerId === RankingsRoute) {
			employerRoutedComponent = <EmployerBracketList mode="ranking" />;
		} else {
			employerRoutedComponent = <EmployerPage employerId={employerId} />;
		}
	} else {
		employerRoutedComponent = <EmployerListPage />;
	}

	return (
		<EmployerRouteContext.Provider value={contextData}>
			{employerRoutedComponent}
		</EmployerRouteContext.Provider>
	);
};

export default EmployerRoute;
