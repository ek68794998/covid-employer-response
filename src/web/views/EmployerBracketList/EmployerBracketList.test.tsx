import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { mockComponent } from "../../../__tests__/TestUtils";

import { AppState } from "../../state/AppState";
import configureStore from "../../state/configureStore";

import { EmployerRouteContext, DefaultContextData } from "../EmployerRoute/EmployerRouteContext";

import EmployerBracketList from "./EmployerBracketList";

jest.mock(
	"../EmployerListItem/EmployerListItem",
	() => mockComponent("EmployerListItem"));

describe("<EmployerList />", () => {
	test("renders without exploding", () => {
		const store: Store<AppState, AnyAction> = configureStore({});

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployerRouteContext.Provider value={DefaultContextData}>
							<EmployerBracketList employers={[]} metric="rating" title="Title" />
						</EmployerRouteContext.Provider>
					</BrowserRouter>
				</Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});
});
