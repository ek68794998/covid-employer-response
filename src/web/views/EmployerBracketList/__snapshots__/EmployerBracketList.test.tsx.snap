// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmployerList /> renders without exploding 1`] = `
<div
  className="EmployerBracketList__Container"
>
  <h2>
    Title
  </h2>
  <div
    className="EmployerBracketList__Grid"
  />
</div>
`;
